### PR TITLE
Add letsencrypt service into nginx-proxy

### LINKED ISSUE #274: can you add Let's Encrypt
Our project is very nice.
It is possible to add the Let's Encrypt Script to get free and automated SSL-certificates for each container?
Look here: https://github.com/letsencrypt/letsencrypt/tree/master/letsencrypt-nginx

That can be the whipped cream. ;)
P.S. Let's Encrypt should be availability for production in  week of November 16, 2015
https://letsencrypt.org/2015/08/07/updated-lets-encrypt-launch-schedule.html

### PR DESCRIPTION
**Pull request by JrCs**
Add transparent integration of letsencrypt service into nginx-proxy.
Only add 40MB in the image size.

Close #274, #300

### CODE REVIEW

**Comment by wader** [comment]
Nice. Maybe cert files update needs to be atomic (per domain?) so that nginx reload because docker event don't happen to see corrupt or missing crt/key file for domains?

**Comment by JrCs** [comment]
@wader it is the case. Reload of nginx and creation/renewal of certs are optimise to only be done if necessary.
You can try my automatic build docker image: jrcs/nginx-proxy-letsencrypt:nginx-proxy

**Comment by PunKeel** [comment]
:+1: for this, but why are you using "LETSENCRYPT_HOST" instead of pulling the list of hosts from the generated configuration ? (That's the whole point of this system :/)

Also, we can't specify one emails per "host" ... Maybe might it use one variable EMAIL per "guest" container ?

**Comment by JrCs** [comment]
@PunKeel LETSENCRYPT_HOST and LETSENCRYPT_EMAIL are used to select the SSL certificates to create.
VIRTUAL_HOST is used by ngnix-proxy to redirect some domains to containers

**Comment by PunKeel** [comment]
This `VIRTUAL_HOST` is useless, right ? (Sorry, forgot to include the link…)

https://github.com/JrCs/docker-nginx-proxy-letsencrypt/commit/670d175ee948a33bb284f2059a3040f6e099867e#diff-04c6e90faac2675aa89e2176d2eec7d8R139

**Comment by JrCs** [comment]
@PunKeel NO the VIRTUAL_HOST is necessary for nginx to redirect domains !
VIRTUAL_XXX variables are use to configure nginx proxy.
LETSENCRYPT_XXX variables are use by letsencrypt service

File: README.md [hunk 1]
```diff
@@ -121,6 +121,28 @@ to establish a connection.  A self-signed or generic cert named `default.crt` an
...
+
+For example
+
+```
+$ docker run -d -p 80:80 -p 443:443 \
+    -e VIRTUAL_HOST="foo.bar.com,bar.com" \
+    -e LETSENCRYPT_HOST="foo.bar.com,bar.com" \
+    -e LETSENCRYPT_EMAIL="foo@bar.com" \
+    -v /var/run/docker.sock:/tmp/docker.sock:ro \
+    jwilder/nginx-proxy
```
Comment (ArthurHoaro) [inline README.md]: Here it should be `...` or `whatever/image`, otherwise it's misleading.

**Comment by atomaka** [comment]
@jcrs Using `LETSENCRYPT_HOST` requires us to configure hosts to use SSL.  Let's Encrypts goal is to have SSL always on. Wouldn't it make more sense to match this goal and have SSL enabled be default for all hosts?  Then, you could just use all hosts registered and create SSL certificates for each.  If for some reason SSL cannot be enabled, perhaps this should be configured per guest with something like `LETS_ENCRYPT_ENABLED=false`.

Awesome work though, I set it up and initial certificates come through. I'll let you know in 90 days if renewals work :)

**Comment by PunKeel** [comment]
:+1: @atomaka ...

**Comment by JrCs** [comment]
@atomaka: LETSENCRYPT_HOST is used to defined the domain for which we want to create certicates with letsencrypt service.
SSL is automatically configured by the nginx-proxy if a certificate exists in the certs directory (certs can be generated by letsencrypt or other service).
So if you have a container with environment variables like:
```pre
VIRTUAL_HOST=www.d1.com, www.d2.com, www.d3.com
LETSENCRYPT_HOST=www.d3.com
```
and you only have a certs for www.d2.com then
* SSL will NOT be enabled for www.d1.com (no certificate)
* SSL will be enabled for www.d2.com (certificate exists in certs directory)
* SSL will be enabled for www.d3.com (certificate will be generate/renew with letsencrypt)

**Comment by PunKeel** [comment]
That still doesn't tell us why `LETSENCRYPT_HOST` is defined on nginx and not on each container.

**Comment by atomaka** [comment]
@PunKeel it is defined per guest container and not on the nginx-proxy container.  The README is a little misleading.

@JrCs I hadn't considered the case where an external certificate was being used. Still, the most common use case seems like it would have `LETSENCRYPT_HOST` matching `VIRTUAL_HOST`` every time and this is a little cumbersome to setup.

File: README.md [hunk 2]
```diff
@@ -121,6 +121,28 @@ to establish a connection.  A self-signed or generic cert named `default.crt` an
...
+```
```
Comment (atomaka) [inline README.md]: This `docker run` command is a misleading.  `VIRTUAL_HOST` and the `LETSENCRYPT_*` environmental variables are set on your guest containers and not on the `jwilder/nginx-proxy` image.

**Comment by ArthurHoaro** [comment]
I've tested https://github.com/JrCs/docker-nginx-proxy-letsencrypt and it works fine - except the README problem.

Possible improvements from user point of view (but not really necessary):

  * define the email once on the proxy container and use it when not specified.
  * if `LETSENCRYPT` is set to `1` (or any variable name/value) on a webapp container, use just use `VIRTUAL_HOST` value instead of copy-pasting it into `LETSENCRYPT_HOST`.

**Comment by Toobian** [comment]
Good job. it's working ! almost... =/
When i try to enable LETSENCRYPT option, i have lot of redirect's error. It does not happen when i use self-signed certificate.

Do you have any idea ?

**Comment by JrCs** [comment]
Closing this PR.
Better to use a dedicated container: [letsencrypt-nginx-proxy-companion](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion)

**Pull request closed without merging**